### PR TITLE
refactor meta

### PR DESCRIPTION
**Pull request by yiming187**

### CODE REVIEW

File: hermes-metaservice/src/main/java/com/ctrip/hermes/metaservice/build/ComponentsConfigurator.java [hunk 1]
```diff
@@ -25,6 +26,9 @@
 	public List<Component> defineComponents() {
 		List<Component> all = new ArrayList<Component>();
 
+		all.addAll(new FxhermesmetadbDatabaseConfigurator().defineComponents());
```
Comment (songtong) [inline hermes-metaservice/src/main/java/com/ctrip/hermes/metaservice/build/ComponentsConfigurator.java]: already defined in the same file

File: hermes-metaservice/src/main/java/com/ctrip/hermes/metaservice/build/ComponentsConfigurator.java [hunk 2]
```diff
@@ -25,6 +26,9 @@
...
+		all.addAll(new FxHermesShardDbDatabaseConfigurator().defineComponents());
```
Comment (songtong) [inline hermes-metaservice/src/main/java/com/ctrip/hermes/metaservice/build/ComponentsConfigurator.java]: same as above

File: hermes-metaservice/src/main/java/com/ctrip/hermes/metaservice/service/DefaultMetaService.java [hunk 3]
```diff
@@ -24,13 +56,279 @@
...
+
+	@Inject
+	protected DatasourceDao m_datasourceDao;
+
+	private Meta m_metaModel;
+
+	private com.ctrip.hermes.meta.entity.Meta m_metaEntity;
+
+	@Override
+	public synchronized com.ctrip.hermes.meta.entity.Meta buildNewMeta() throws DalException {
```
Comment (songtong) [inline hermes-metaservice/src/main/java/com/ctrip/hermes/metaservice/service/DefaultMetaService.java]: need to notify meta-server by zk

File: hermes-metaservice/src/main/java/com/ctrip/hermes/metaservice/service/DefaultMetaService.java [hunk 4]
```diff
@@ -24,13 +56,279 @@
...
+			entity.addPartition(p);
+		}
+		List<com.ctrip.hermes.meta.entity.Producer> producers = findProducers(model);
+		for (com.ctrip.hermes.meta.entity.Producer p : producers) {
+			entity.addProducer(p);
+		}
+		return entity;
+	}
+
+	public com.ctrip.hermes.metaservice.model.Meta getMetaModel() throws DalException {
```
Comment (songtong) [inline hermes-metaservice/src/main/java/com/ctrip/hermes/metaservice/service/DefaultMetaService.java]: Does getMetaEntity() below need to throws exceptions?

File: hermes-metaservice/src/main/java/com/ctrip/hermes/metaservice/service/DefaultMetaService.java [hunk 5]
```diff
@@ -24,13 +56,279 @@
...
+				refreshMeta();
+			} catch (DalException e) {
+				m_logger.warn("get meta entity failed", e);
+				return new com.ctrip.hermes.meta.entity.Meta();
+			}
+		}
+		return m_metaEntity;
+	}
+
+	protected boolean isMetaUpdated() {
```
Comment (songtong) [inline hermes-metaservice/src/main/java/com/ctrip/hermes/metaservice/service/DefaultMetaService.java]: Should isMetaUpdated() do update meta?

File: hermes-metaservice/src/main/java/com/ctrip/hermes/metaservice/service/DefaultPortalMetaService.java [hunk 6]
```diff
@@ -33,6 +27,8 @@
 import com.ctrip.hermes.meta.entity.Property;
 import com.ctrip.hermes.meta.entity.Storage;
 import com.ctrip.hermes.meta.entity.Topic;
+import com.ctrip.hermes.metaservice.model.DatasourceEntity;
+import com.ctrip.hermes.metaservice.model.TopicEntity;
 
 @Named(type = PortalMetaService.class, value = DefaultPortalMetaService.ID)
 public class DefaultPortalMetaService extends DefaultMetaService implements PortalMetaService, Initializable {
```
Comment (songtong) [inline hermes-metaservice/src/main/java/com/ctrip/hermes/metaservice/service/DefaultPortalMetaService.java]: 之前的DefaultPortalMetaService是有缓存的，现在没有了，会对性能有影响吗？

Comment (yiming187) [inline hermes-metaservice/src/main/java/com/ctrip/hermes/metaservice/service/DefaultPortalMetaService.java]: Cache根据需要可以以后加，先保证逻辑正确性。

File: hermes-metaserver/src/main/java/com/ctrip/hermes/metaserver/event/impl/BaseMetaChangedEventHandler.java [hunk 7]
```diff
@@ -44,7 +44,7 @@ public EventType eventType() {
 
 	@Override
 	protected void processEvent(Event event) throws Exception {
-		Meta baseMeta = m_metaService.findLatestMeta();
+		Meta baseMeta = m_metaService.getMetaEntity();
```
Comment (leoliang) [inline hermes-metaserver/src/main/java/com/ctrip/hermes/metaserver/event/impl/BaseMetaChangedEventHandler.java]: 应该调用refreshMeta()

File: hermes-metaserver/src/main/java/com/ctrip/hermes/metaserver/event/impl/LeaderInitEventHandler.java [hunk 8]
```diff
@@ -203,7 +203,7 @@ private Long loadAndAddBaseMetaWatcher(Watcher watcher) throws Exception {
 	}
 
 	private Meta loadBaseMeta() throws Exception {
-		return m_metaService.findLatestMeta();
+		return m_metaService.getMetaEntity();
```
Comment (leoliang) [inline hermes-metaserver/src/main/java/com/ctrip/hermes/metaserver/event/impl/LeaderInitEventHandler.java]: 应该调用refreshMeta()

**Pull request merged**